torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -509,28 +510,26 @@ void ProcessGroupAgent::pollTimedOutRPCs() {
       return;
     }
 
-    const auto timedOutFutures = processTimedOutFutures();
+    const auto timedOutFutures = processTimedOutFutures(std::move(lock));
```

Comment:
Why do we need to pass in the lock into this function? We're already holding the lock when this function is called (see like 488)

Reply:
I unlock the lock at the end of this function.

Reply:
I guess the question is why do we pass the lock to `processTimedOutFutures` instead of holding the lock for it in this function? Doing `unlock` and `notify_all` in this function does not seem to introduce any perf penalty but would make the code more readable?

Reply:
Since we have to acquire lock before testing rpcRunning_, so we don't want to release the lock and re-acquire the lock in the function.

Unless we find a way to not hholding the lock while testing rpcRunning, we can even put all the code about mutating future map and future timeout map into the helper function, `processTimedOutFutures`.

Exactly, notify the cond var as soon as the lock can be released is just for redability.

Reply:
To chime in, I think the code would be more readable if we didn't pass the lock to the function, but instead manually signal the condvar after, i.e.

```
    const auto timedOutFutures = processTimedOutFutures();
    lock.unlock();
    futureCV_.notify_all();
```

But if others disagree, this is just a weak preference. :) My feeling here is that we want it to be easy to reason about the correctness of the locking here, and std::move()ing the unique_lock is usually best when we have no choice.

Reply:
@jjlilley 
In the latest version, it's already the same as what you have in mind.